File: solutions/ocaml/155-ransom-note.ml
(*
  Problem 155: Ransom Note (LeetCode 383)
  Difficulty: Easy
  Language: OCaml
*)
let can_construct note magazine =
  let freq = Hashtbl.create 26 in
  String.iter (fun ch ->
    let c = try Hashtbl.find freq ch with Not_found -> 0 in
    Hashtbl.replace freq ch (c + 1)) magazine;
  String.fold_left (fun ok ch ->
    let c = try Hashtbl.find freq ch with Not_found -> 0 in
    Hashtbl.replace freq ch (c - 1); ok && c > 0) true note

File: solutions/rust/155-ransom-note.rs
/**
 * Problem 155: Ransom Note (LeetCode 383)
 * Difficulty: Easy
 * Language: Rust
 */
fn can_construct(ransom: &str, magazine: &str) -> bool {
    let freq = magazine.chars().fold(std::collections::HashMap::new(), |mut m, c| {
        *m.entry(c).or_insert(0) += 1; m
    });
    ransom.chars().fold((freq, true), |(mut m, ok), c| {
        let cnt = m.entry(c).or_insert(0);
        *cnt -= 1;
        (m, ok && *cnt >= 0)
    }).1
}
